# HG changeset patch
# User chrisphi
# Date 1523978139 14400
#      Tue Apr 17 11:15:39 2018 -0400
# Node ID 71093353a556fce07fb430000adfcbfa20a1bbb2
# Parent  fcff2daa6b1eaae1c69bd930e590d2b72121795f
8201509: Zero : S390x (S390 and not _LP64) atomic_copy64 inline assembler is wrong.
Summary: In os_linux_zero.hpp the inline assembler for the S390 (S390 and not _LP64) has src and dst reversed thereby corrupting data.
Reviewed-by: Duke
Contributed-by: MBalao

diff --git openjdk/hotspot/src/os_cpu/linux_zero/vm/os_linux_zero.hpp openjdk/hotspot/src/os_cpu/linux_zero/vm/os_linux_zero.hpp
--- openjdk/hotspot/src/os_cpu/linux_zero/vm/os_linux_zero.hpp
+++ openjdk/hotspot/src/os_cpu/linux_zero/vm/os_linux_zero.hpp
@@ -50,10 +50,10 @@
                   : "Q"(*(volatile long*)src));
 #elif defined(S390) && !defined(_LP64)
     double tmp;
-    asm volatile ("ld  %0, 0(%1)\n"
-                  "std %0, 0(%2)\n"
-                  : "=r"(tmp)
-                  : "a"(src), "a"(dst));
+    asm volatile ("ld  %0, %2\n"
+                  "std %0, %1\n"
+                  : "=&f"(tmp), "=Q"(*(volatile double*)dst)
+                  : "Q"(*(volatile double*)src));
 #else
     *(jlong *) dst = *(const jlong *) src;
 #endif
